### PR TITLE
Vulkan on Windows

### LINKED ISSUE #59519: SPIR-V codegen fails when building on Windows with Vulkan
## To Reproduce

1. Install `VULKAN_SDK` `1.2.176.1`
2. Install `miniconda` `3.8.10`
3. Install build deps `conda install -y astunparse numpy ninja pyyaml mkl mkl-include setuptools cmake cffi typing_extensions future six requests dataclasses`
4. clone commit `1d9c1c` and `cd  pytorch`
5. build wheel:
```pwsh
$env:USE_VULKAN=1
$env:USE_VULKAN_WRAPPER=0
python setup.py bdist_wheel
```

Build errors because `tools` module is missing. It seems `pytorch` is not added to `PYTHONPATH` in [VulkanCodegen.cmake ](https://github.com/pytorch/pytorch/blob/master/cmake/VulkanCodegen.cmake#L58).
```
Traceback (most recent call last):
  File "D:/a/pytorch-etna/pytorch-etna/pytorch/cmake/../aten/src/ATen/gen_vulkan_spv.py", line 8, in <module>
    from tools.codegen.code_template import CodeTemplate
ModuleNotFoundError: No module named 'tools'
CMake Error at cmake/VulkanCodegen.cmake:72 (message):
  Failed to gen spv.h and spv.cpp with precompiled shaders for Vulkan backend
Call Stack (most recent call first):
  caffe2/CMakeLists.txt:6 (include)
``` 

Build summary:
```
-- ******** Summary ********
--   CMake version         : 3.19.6
--   CMake command         : C:/Miniconda/envs/test/Library/bin/cmake.exe
--   System                : Windows
--   C++ compiler          : C:/Program Files (x86)/Microsoft Visual Studio/2019/Enterprise/VC/Tools/MSVC/14.29.30037/bin/Hostx64/x64/cl.exe
--   C++ compiler version  : 19.29.30037.0
--   CXX flags             : /DWIN32 /D_WINDOWS /GR /EHsc /w /bigobj -DUSE_PTHREADPOOL -openmp:experimental
--   Build type            : Release
--   Compile definitions   : WIN32_LEAN_AND_MEAN;ONNX_ML=1;ONNXIFI_ENABLE_EXT=1
--   CMAKE_PREFIX_PATH     : C:\Miniconda\envs\test\Lib\site-packages
--   CMAKE_INSTALL_PREFIX  : D:/a/pytorch-etna/pytorch-etna/pytorch/torch
--   CMAKE_MODULE_PATH     : D:/a/pytorch-etna/pytorch-etna/pytorch/cmake/Modules;D:/a/pytorch-etna/pytorch-etna/pytorch/cmake/public/../Modules_CUDA_fix
```

## Expected behavior

Build succeeds.

## Environment
```
OS: Microsoft Windows Server 2019 Datacenter
GCC version: Could not collect
Clang version: Could not collect
CMake version: version 3.19.6
Libc version: N/A

Python version: 3.8 (64-bit runtime)
Python platform: Windows-10-10.0.17763-SP0
Is CUDA available: N/A
CUDA runtime version: Could not collect
GPU models and configuration: Could not collect
Nvidia driver version: Could not collect
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A

Versions of relevant libraries:
[pip3] numpy==1.20.2
[conda] blas                      1.0                         mkl  
[conda] mkl                       2021.2.0           haa95532_296  
[conda] mkl-include               2021.2.0           haa95532_296  
[conda] mkl-service               2.3.0            py38h2bbff1b_1  
[conda] mkl_fft                   1.3.0            py38h277e83a_2  
[conda] mkl_random                1.2.1            py38hf11a4ad_2  
[conda] numpy                     1.20.2           py38ha4e8547_0  
[conda] numpy-base                1.20.2           py38hc2deb75_0
```

This is on a `windows-2019` github action runner.  The equivalent error happens on a `Windows 10 Pro` desktop.

## Additional context

This issue seems to be a subset or specific version of #55879.

cc @malfet @seemethere @walterddr @peterjc123 @mszhanyi @skyline75489 @nbcsm

### PR DESCRIPTION
**Pull request by ahirner**
Fixes #59519 and partly #55879.
Credits/continued from: @skyline75489's [diff](https://github.com/pytorch/pytorch/compare/master...skyline75489:chesterliu/dev/vulkan-on-win?expand=1)

Host:
```
Windows: 10.0.19042
Vulkan SDK: 1.2.176.1
CPU: i5-8250U
Build Tools 2019: 16.8.31005.135 
```
The following operations succeeded:
```
Python 3.8.10 (default, May 19 2021, 13:12:57) [MSC v.1916 64 bit (AMD64)] :: Anaconda, Inc. on win32
Type "help", "copyright", "credits" or "license" for more information.
>>> import torch
>>> torch.is_vulkan_available()
True
>>> t1 = torch.randn(2,3,224,224)
>>> t2 = torch.randn(2,3,224,224)
>>> t3 = t1 + t2
>>> t3_v = t1.to(device='vulkan') + t2.to(device='vulkan')
>>> t3_c = t3_v.cpu()
>>> torch.allclose(t3, t3_c)
True
>>> t3_v.device
device(type='vulkan')
```

### CODE REVIEW

File: aten/src/ATen/native/vulkan/Vulkan.cpp [hunk 1]
```diff
@@ -24,8 +24,10 @@
 #include <iostream>
 #include <numeric>
 #include <stdio.h>
+#ifdef _MSC_VER
+#else
```
Comment (malfet) [inline aten/src/ATen/native/vulkan/Vulkan.cpp:28]: Nit
```suggestion
#ifndef _MSC_VER
```

File: aten/src/ATen/native/vulkan/api/Context.cpp [hunk 2]
```diff
@@ -42,8 +42,12 @@ VkDevice create_device(
       nullptr,
       &device_extension_properties_count,
       device_extension_properties.data()));
-
+  #ifdef _MSC_VER
+  // MSVC does not support VLAs (variable length arrays).
+  std::vector<const char*> requested_device_extensions{
+  #else
   constexpr const char* const requested_device_extensions[]{
+  #endif
```
Comment (malfet) [inline aten/src/ATen/native/vulkan/api/Context.cpp:50]: Let compiler figure out the best type (I suspect it would be [std::initializer_list](https://en.cppreference.com/w/cpp/utility/initializer_list))
```suggestion
  constexpr auto requested_device_extensions = {
  #endif
```

File: aten/src/ATen/native/vulkan/api/Context.h [hunk 3]
```diff
@@ -135,6 +135,19 @@ inline void bind(
...
+    const Shader::Layout::Signature&,
+    const Shader::Descriptor&,
+    const Shader::WorkGroup&);
+
+// Forward declaration
+void dispatch_epilogue(
+    Command::Buffer&,
+    const Descriptor::Set&,
+    const Shader::WorkGroup&);
+
```
Comment (malfet) [inline aten/src/ATen/native/vulkan/api/Context.h:150]: Can you please explain why this is needed, especially in context of Windows ?

File: aten/src/ATen/native/vulkan/api/Resource.cpp [hunk 4]
```diff
@@ -87,7 +87,7 @@ void release_image(const Resource::Image& image) {
 
 } // namespace
 
-void* map(
+void* _map(
```
Comment (malfet) [inline aten/src/ATen/native/vulkan/api/Resource.cpp:90]: Why this change is needed?

File: c10/util/flat_hash_map.h [hunk 5]
```diff
@@ -233,12 +233,14 @@ template <
     typename T,
     typename FindKey,
     typename ArgumentHash,
-    typename Hasher,
+    typename DetailHasher,
```
Comment (malfet) [inline c10/util/flat_hash_map.h:236]: I'm not sure I understand how this change is related to enabling Vulka on Windows

**Review by malfet — CHANGES_REQUESTED** [review]
Can you please get rid of unrelated changes (or explain why they are necessary)
Or perhaps split your PR in several smaller ones, namely:
 - One that get rid of variable-length-array  and init structure fields by name extensions
 -  One that replaces `os.popen("find")` with `os.walk` and updates CMakeLists to have correct Windows invocations

Replacing `Hasher` with `DetailHasher` sounds strange and completely unrelated

File: c10/util/flat_hash_map.h (hunk 5 above)
Comment (skyline75489) [reply c10/util/flat_hash_map.h:236]: This is related to the a VS bug(feature?). Link [here](https://developercommunity.visualstudio.com/t/nested-templates-causing-compilation-error-when-th/1397858?from=email#T-ND1399286)

**Comment by xsacha** [comment]
For some reason, after merging this in to master, I get the following CMake error:
```
CMake Error: Error required internal CMake variable not set, cmake may not be built correctly.
Missing variable is:
CMAKE_ASM_COMPILE_OBJECT
```

To solve it, I had to add 'ASM' to the project line in CMakeLists.txt

**Comment by ahirner** [comment]
Thanks @xsacha for the notice. This is a nice heads up! 
I tried to make progress with a smaller scoped and rebased version too. Unfortunately our windows runner OOMed so far... and hence time to do anything went up unexpectedly.

File: c10/util/flat_hash_map.h (hunk 5 above)
Comment (ahirner) [reply c10/util/flat_hash_map.h:236]: I couldn't find a way to control `/permissive` per file with a #pragma either. Does another trick come to anyone's mind here?
